src/compiler/scala/tools/nsc/transform/Erasure.scala

```diff
@@ -1216,14 +1216,24 @@ abstract class Erasure extends InfoTransform
           Match(Typed(selector, TypeTree(selector.tpe)), cases)
 
         case Literal(ct) =>
-          // We remove the original tree attachments in pre-easure to free up memory
+          // We remove the original tree attachments in pre-erasure to free up memory
           val cleanLiteral = tree.removeAttachment[OriginalTreeAttachment]
 
           if (ct.tag == ClazzTag && ct.typeValue.typeSymbol != definitions.UnitClass) {
-            val erased = ct.typeValue.dealiasWiden match {
-              case tr @ TypeRef(_, clazz, _) if clazz.isDerivedValueClass => scalaErasure.eraseNormalClassRef(tr)
-              case tpe => specialScalaErasure(tpe)
-            }
+            val typeValue = ct.typeValue
+
+            /* scala/bug#10551: Look at the symbol of the type to see if it's
+             * a derived value class, so classOf[it] doesn't get erased to the
+             * underlying type.
+             *
+             * Derived value classes can't be nested inside of other (non-static)
+             * classes, so rather than fiddling around with the erasure map, just
+             * make a vanilla `TypeRef` to the underlying symbol. */
+            val erased =
+              if (typeValue.typeSymbol.isDerivedValueClass)
+                typeRef(NoPrefix, typeValue.typeSymbol, Nil)
+              else specialScalaErasure(typeValue)
```

Comment:
What happened to dealiasWiden?

Reply:
It went the way of the unpartested dodo. Good catch. I wonder how I could test that.

Reply:
I poked at it a bit, and I'm not experienced enough to come up with a reason why the `dealiasWiden` should matter (after this patch).

I'm not sure what it fixed before, so I've put it back.

Reply:
Bikeshedding a bit.

We could just unconditionally call `specialScalaErasure` and check for the value classity by seeing if we get an `ErasedValueType` back:

```
scala> erasure.specialScalaErasure(typeOf[C]) match { case ErasedValueType(valueClass, _) => valueClass.tpeHK; case tp => tp }
res5: $r.intp.global.Type = C
```

We can also use `valueClassSymbol.tpeHK` rather than `typeRef(NoPrefix, ..., Nil)`. That is cached internally so we save a few cycles.

Reply:
Or just use:

```
scala> erasure.javaErasure(typeOf[C])
res6: $r.intp.global.erasure.global.Type = C
```

I'd suggest to put in the different options and assert they give the same results (as I think they would).

Reply:
Poking at those suggestions (thanks!), a corner case arises (arrayses?):

My (current) version, my version with `.tpeHK`, and matching on `specialScalaErasure`: 
```scala
scala> :pa -raw
// Entering paste mode (ctrl-D to finish)

class A[T](val a: T) extends AnyVal

// Exiting paste mode, now interpreting.

scala> classOf[Array[A[_]]]
res0: Class[Array[A[_]]] = class [Ljava.lang.Object;

scala> classOf[Array[A[Int]]]
res1: Class[Array[A[Int]]] = class [LA;
```

Using `javaErasure`:
```scala
scala> :pa -raw
// Entering paste mode (ctrl-D to finish)

class A[T](val a: T) extends AnyVal

// Exiting paste mode, now interpreting.


scala> classOf[Array[A[_]]]
res0: Class[Array[A[_]]] = class [LA;

scala> classOf[Array[A[Int]]]
res1: Class[Array[A[Int]]] = class [LA;
```

But... that's not the end of the array problems...